src/net_processing.cpp

```diff
@@ -1546,56 +1546,60 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
     return ret;
 }
 
-bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) const
+CNodeStateStats PeerManagerImpl::GetNodeStateStats(NodeId nodeid) const
 {
+    CNodeStateStats stats;
     {
         LOCK(cs_main);
         const CNodeState* state = State(nodeid);
-        if (state == nullptr)
-            return false;
-        stats.nSyncHeight = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
-        stats.nCommonHeight = state->pindexLastCommonBlock ? state->pindexLastCommonBlock->nHeight : -1;
+        if (state == nullptr) {
+            // Return member-initialized sensible defaults for the fields where
+            // it makes sense and simplifies the code.
+            return stats;
+        }
+        if (state->pindexBestKnownBlock) {
+            stats.nSyncHeight = state->pindexBestKnownBlock->nHeight;
+        }
+        if (state->pindexLastCommonBlock) {
+            stats.nCommonHeight = state->pindexLastCommonBlock->nHeight;
+        }
         for (const QueuedBlock& queue : state->vBlocksInFlight) {
-            if (queue.pindex)
+            if (queue.pindex) {
                 stats.vHeightInFlight.push_back(queue.pindex->nHeight);
+            }
         }
     }
-
-    PeerRef peer = GetPeerRef(nodeid);
-    if (peer == nullptr) return false;
+    const PeerRef peer = GetPeerRef(nodeid);
+    if (peer == nullptr) {
+        // Peer is still being set up; return the stats that we do have.
+        return stats;
+    }
     stats.their_services = peer->m_their_services;
-    stats.m_starting_height = peer->m_starting_height;
+    if (peer->m_starting_height != -1) {
```

Comment:
This is handling the two ways our code looks at `starting_height` unknown or something, which we create in b4bf7611d718e361dc7dea849550b3728da79136. Why not having one way? Switch to a new default if you wish, but having two different ones seems suboptimal.

Reply:
I didn't do that yet as it would extend the refactoring from `CNodeStateStats` into the `Peer` class to change `m_starting_height` from `std::atomic` to `std::optional` with a mutex, but might still do it once this PR becomes refactoring-only (the other changes have been unbundled).

Reply:
> I didn't do that yet as it would extend the refactoring from `CNodeStateStats` into the `Peer` class to change `m_starting_height` from `std::atomic` to `std::optional` with a mutex, but might still do it once this PR becomes refactoring-only (the other changes have been unbundled).

Started doing this in latest (WIP) push.